### PR TITLE
Fast sparse add.

### PR DESCRIPTION
**Pull request by dr-robertk**
This PR improves performance of flow by introducing an improved sparse add method for AutoDiffMatrix. When two SparseMatrix have the same sparsity pattern, we can simply add all matrix entries. Performance improvement of close to 10% has been discovered with SPE9 and Norne. 

This PR already includes #578. I will rebase once the other PR is merged.

### CODE REVIEW

**Comment by totto82** [comment]
I have tested this on Model 2 and the result are good. Some deviations in results can be observed but the deviations can be explained by differences in the time-stepping.  
The total run time is reduced and also the number of linear and Newton iterations.

**Comment by atgeirr** [comment]
> I have tested this on Model 2 and the result are good

Excellent! With that, I will merge the first PR #578, and review the code here when rebased.

**Comment by dr-robertk** [comment]
Rebased.

File: opm/autodiff/AutoDiffBlock.hpp [hunk 1]
```diff
@@ -374,14 +378,31 @@ namespace Opm
                     jac[block] = M( D3.rows(), jac_[block].cols() );
                 }
                 else if( jac_[block].nonZeros() == 0 ) {
-                    jac[block] =  D3 * ( D1*rhs.jac_[block]) * (-1.0);
+#pragma omp critical
```
Comment (atgeirr) [inline opm/autodiff/AutoDiffBlock.hpp]: To be sure I understand this: the #pragma applies to the following if() only, correct?

File: opm/autodiff/AutoDiffMatrix.hpp [hunk 2]
```diff
@@ -264,7 +264,12 @@ namespace Opm
 
         AutoDiffMatrix& operator+=(const AutoDiffMatrix& rhs)
         {
-            *this = *this + rhs;
+            if( type_ == Sparse && rhs.type_ == Sparse )
+            {
+                fastSparseAdd( sparse_, rhs.sparse_ );
+            }
+            else
+                *this = *this + rhs;
```
Comment (atgeirr) [inline opm/autodiff/AutoDiffMatrix.hpp]: Please use braces around the else part.

File: opm/autodiff/fastSparseOperations.hpp [hunk 3]
```diff
@@ -177,7 +177,89 @@ inline void fastSparseDiagProduct(const Eigen::SparseMatrix<double>& lhs,
...
+  // check complete sparsity pattern
+  if( equal )
+  {
+    typedef typename Eigen::internal::remove_all<Lhs>::type::Index Index;
+    const Index nnz = lhs.nonZeros();
 
+    const Index* rhsJ = rhs.innerIndexPtr();
+    const Index* lhsJ = lhs.innerIndexPtr();
+
+    for(Index i=0; i<nnz; ++i )
```
Comment (atgeirr) [inline opm/autodiff/fastSparseOperations.hpp]: Is this sufficient? It checks that all entries have the same inner index (row number for a column major matrix), but it is possible that two matrices have matching inner indices while not the same pattern. For example:
```
1 0 0
0 1 1
0 0 1
```
has the same inner pattern (assuming column major) as
```
1 0 0
1 1 0
0 0 1
```

Comment (dr-robertk) [inline opm/autodiff/fastSparseOperations.hpp]: In that case inner indices of matrix one would be 0,1,2,2 and the second matrix would have 0,0,1,2 which is not the same, thus not fast sparse add.

Comment (atgeirr) [inline opm/autodiff/fastSparseOperations.hpp]: I think the inner indices would be 0, 1, 1, 2 for both matrices if they are column major, right?

**Comment by atgeirr** [comment]
Can we try to convince Travis to rebuild this? Must we close and reopen it?

**Comment by dr-robertk** [comment]
Thanks for pointing out the equalSparsityPattern problem. That should be fixed now. I included the outerIndices. In addition I reverted some changes in AutoDiffBlock that did not work.

**Comment by joakim-hove** [comment]
With normal user credentials I think that is the only way.
16. feb. 2016 11:13 skrev "Atgeirr Flø Rasmussen" <notifications@github.com>
følgende:

**Comment by atgeirr** [comment]
Will close and reopen to trigger Travis.

**Comment by dr-robertk** [comment]
@bska: Why can't one use bitwise operators for bool? My guess is that these are correctly overloaded.

**Comment by dr-robertk** [comment]
@atgeirr: Checking both arrays in one loop might allow for some vectorization.

File: opm/autodiff/AutoDiffMatrix.hpp (hunk 2 above)
Comment (dr-robertk) [inline opm/autodiff/AutoDiffMatrix.hpp]: Fixed.

**Comment by bska** [comment]
> @bska: Why can't one use bitwise operators for bool?

Bitwise operators are defined entirely in terms of "promoted integral types".  The type `bool` promotes to `int` which is unsafe for bitwise operations (`unsigned int` is safe, however).

> My guess is that these are correctly overloaded.

Code should be correct and work for the right reasons.  There should be no need to guess anywhere.

**Comment by dr-robertk** [comment]
I have addressed all nitpicks. Performance  improvement was observed for SPE9 (ca 10%) and Norne (ca 5%)

**Comment by atgeirr** [comment]
> Checking both arrays in one loop might allow for some vectorization.

I would think that the memory must be continuous for that? Anyway, thanks a lot for the effort, will merge!

**Pull request merged**